### ISSUE TITLE
OS version in the library description is ignored

### ISSUE THREAD

**Issue by 2frei** [toolbox4minecraft/amidst#171]
Tried with Amidst in jar version with version : 3.6, 3.7, 4.0, 4.1, 4.2.beta1 and 4.2.beta2 and I have always the same result : it works with MC1.9.4 or MC1.8.9 but crashes with MC1.5.2 or MC1.4.7 :

```
Amidst has encountered an uncaught exception on the thread Thread[Thread-6,1,main]
[info] Amidst v4.1
[info] Current system time: 2016-05-17 15:33:05.537
[info] System.getProperty("os.name") == 'Mac OS X'
[info] System.getProperty("os.version") == '10.8.5'
[info] System.getProperty("os.arch") == 'x86_64'
[info] System.getProperty("java.version") == '1.8.0_73'
[info] System.getProperty("java.vendor") == 'Oracle Corporation'
[info] System.getProperty("sun.arch.data.model") == '64'
[info] Enabling OpenGL.
[info] Biome profile activated: default
[info] using '.minecraft' directory at: '/Users/2frei/Library/Application Support/minecraft', libraries: '/Users/2frei/Library/Application Support/minecraft/libraries'
[info] using biome profiles at: 'biome'
[info] Scanning for profiles.
[info] Successfully loaded profile list.
[info] Beginning latest version list load.
[info] Attempting to download remote version list…
[warning] cannot find valid profile directory for launcher profile 'MapWriter': /Volumes/Mac_Docs/2frei/Jeux/Minecraft/Profiles/MapWriter
[info] Successfully loaded version list. URL: https://launchermeta.mojang.com/mc/game/version_manifest.json
[warning] cannot find valid version directory for launcher profile 'Cubenis'
[warning] cannot find valid version directory for launcher profile 'Epicube'
[info] using minecraft launcher profile 'test' with versionId '1.5.2'
[info] Loading libraries.
[info] Found library: /Users/2frei/Library/Application Support/minecraft/libraries/net/minecraft/launchwrapper/1.5/launchwrapper-1.5.jar
[info] Found library: /Users/2frei/Library/Application Support/minecraft/libraries/net/sf/jopt-simple/jopt-simple/4.5/jopt-simple-4.5.jar
[info] Found library: /Users/2frei/Library/Application Support/minecraft/libraries/org/ow2/asm/asm-all/4.1/asm-all-4.1.jar
[info] Found library: /Users/2frei/Library/Application Support/minecraft/libraries/net/java/jinput/jinput/2.0.5/jinput-2.0.5.jar
[info] Found library: /Users/2frei/Library/Application Support/minecraft/libraries/net/java/jutils/jutils/1.0.0/jutils-1.0.0.jar
[info] Skipping library: org.lwjgl.lwjgl:lwjgl:2.9.0
[info] Skipping library: org.lwjgl.lwjgl:lwjgl_util:2.9.0
[info] Skipping library: org.lwjgl.lwjgl:lwjgl-platform:2.9.0
[warning] Failed attempt to load library at: /Users/2frei/Library/Application Support/minecraft/libraries/org/lwjgl/lwjgl/lwjgl/2.9.1-nightly-20130708-debug3
[info] Skipping library: org.lwjgl.lwjgl:lwjgl:2.9.1-nightly-20130708-debug3
[warning] Failed attempt to load library at: /Users/2frei/Library/Application Support/minecraft/libraries/org/lwjgl/lwjgl/lwjgl_util/2.9.1-nightly-20130708-debug3
[info] Skipping library: org.lwjgl.lwjgl:lwjgl_util:2.9.1-nightly-20130708-debug3
[warning] Failed attempt to load library at: /Users/2frei/Library/Application Support/minecraft/libraries/org/lwjgl/lwjgl/lwjgl-platform/2.9.1-nightly-20130708-debug3
[info] Skipping library: org.lwjgl.lwjgl:lwjgl-platform:2.9.1-nightly-20130708-debug3
[info] Found library: /Users/2frei/Library/Application Support/minecraft/libraries/net/java/jinput/jinput-platform/2.0.5/jinput-platform-2.0.5-natives-osx.jar
[info] Found library: /Users/2frei/Library/Application Support/minecraft/libraries/net/java/jinput/jinput-platform/2.0.5/jinput-platform-2.0.5-natives-osx.jar
[crash] Amidst has encounted an uncaught exception on the thread Thread[Thread-6,1,main]
[crash] java.lang.NoClassDefFoundError: org/lwjgl/LWJGLException
    at java.lang.Class.getDeclaredFields0(Native Method)
    at java.lang.Class.privateGetDeclaredFields(Class.java:2583)
    at java.lang.Class.getDeclaredFields(Class.java:1916)
    at amidst.mojangapi.minecraftinterface.RecognisedVersion.getMainClassFields(RecognisedVersion.java:133)
    at amidst.mojangapi.minecraftinterface.RecognisedVersion.generateMagicString(RecognisedVersion.java:126)
    at amidst.mojangapi.minecraftinterface.RecognisedVersion.from(RecognisedVersion.java:120)
    at amidst.mojangapi.minecraftinterface.local.LocalMinecraftInterfaceBuilder.create(LocalMinecraftInterfaceBuilder.java:34)
    at amidst.mojangapi.file.directory.VersionDirectory.createLocalMinecraftInterface(VersionDirectory.java:106)
    at amidst.mojangapi.MojangApi.set(MojangApi.java:74)
    at amidst.gui.profileselect.LocalProfileComponent.tryLoad(LocalProfileComponent.java:82)
    at amidst.threading.WorkerExecutor.lambda$run$11(WorkerExecutor.java:69)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
    at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.ClassNotFoundException: org.lwjgl.LWJGLException
    at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
    at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
    at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
    ... 14 more
```

**Comment by stefandollase** [comment]
Did you run the failing Minecraft versions at least once? By run, I mean you get to the title screen which contains the Singleplayer and Multiplayer buttons.

**Comment by 2frei** [comment]
Yes, for each MC versions I created a solo game, saved it, and then quit MC.
Tried also with MC 1.6.4, and amidst works correctly.
Tried also with MC 1.4.7, and amidst crashes.

But I found something interresting, there is a problem between MC launcher 1.6.61 (I do not tried with other launchers) and MC versions lower than 1.6 :

In the launcher, I create a profile lower than 1.6 (1.5.2 for exemple) and then I redefine in the profile editor an other game directory.
Then I start MC and created a solo game, cut just one tree, save my game and quit MC.
After that I reopen MC and my solo game have been disapeared.

I open my file explorer and the world files are correctly stored in the new directory.
Then I'm going into the original saved game directory and I find a second directory corresponding to my world, but this one is really badly configured.
I suppressed this directory and copy my world save from the new directory to the original.
I start MC and it's magic : my solo world is new exists.
I open it and play again before save it and quit MC.
Then with my explorer, in the new directory, files dates correspond to me playing time. But in the original MC directory, files dates are wrongs.

Si I deduce that MC loads game from the original saved games directory, but stores it in the new directory. With MC version higher or equal than 1.6, this problem doesn't exist.

I don't know if there is a relation between this fact and my problem with the use of MC < 1.6 but I find this coincidence really strange.
So I created a new profile in 1.5.2 with the original game directory, created a world, played just one minute, saved it and then quit, but amidst crashed again.

PS : I know there are a lot of english faults ... I'm sorry and hope that you will understand what i tried to talk about ... I'm french ^^

**Comment by stefandollase** [comment]
Don't worry about your english, to me it seems to be fine :-) I think I understood everything. However, I am also not a native english speaker.

I just tested the following: Use Amidst v4.1-beta2 with the Minecraft launcher 1.6.61 and a Minecraft profile with the Minecraft version 1.5.2 and a separate gamedir. I started the profile once, as described above and then I used Amidst with the same profile to generate a random map. It worked for me.

You always talk about a save game. Can you try to generate a random map, so we can see whether the issue also exists there?

Did you do anything non-standard to your Minecraft installation? I don't know how the directory structure should look like on OS X, but at a first glance it seems to be fine. However, your description about the Minecraft launcher sounds pretty messed up.

For now, I think this is an issue with the Minecraft launcher: You said you launched the Minecraft profile at least once, but the crash report above says that it is unable to load several libraries. These libraries should have been downloaded by the Minecraft launcher, because you ran the Minecraft profile.

Maybe, you can find or report the issues with the Minecraft Launcher in Mojangs bug tracker, [here](https://bugs.mojang.com/browse/MCL)?

I will wait for your reply before I further investigate the issue.

**Comment by BigAlanM** [comment]
I just changed my Amidstv4.2-beta2 from PCMCv1.9.4 to PCMCv1.4.2 and loaded a saved world with no problems.
Note, however, that I have separate save folders for worlds made with different MC versions.

Worked with PCMCv1.5.2 also.

**Comment by 2frei** [comment]
> You always talk about a save game. Can you try to generate a random map, so we can see whether the issue also exists there?

No, each time I test a new MC version, I first generate a new world, and then I talk about it as the saved world … sorry to not be clear.

If amidst works with you in 1.5.2, then I will perform other tests, and first I will make some cleaning with my profiles and my launcher … yeah I know I performed some cleaning and changes in my directory with my explorer, and the 3 invalid profiles have not been used for my performed test.

All my profiles are pure vanilla except one named BTW that used a properly modded jar with Better Than Wolves mod. And yes I performed the corrected operations with this jar (created a folder '1.5.2-btw' with MC Patcher 5.0.3.

Nevertheless, I needed to modified my .json like this :
[[1.5.2][GUIDE] How-to install BTW w/ MCLauncher & MCPatc](http://www.sargunster.com/btwforum/viewtopic.php?f=9&t=8506&start=20)
maybe this cause a problem in my launcher

**Comment by 2frei** [comment]
I deleted everything in the .minecraft folder
I launched MC to DL a proper version of the launcher
I created one profile and just modified the use of version 1.5.2
I launched MC, generated a new world and just cut a tree
I saved and quitted MC
Start amidst and it crashed (see below).

Maybe it's due to OSX, I will try with java 1.8.0_91

`[info] Amidst v4.2-beta2
[info] Current system time: 2016-05-19 18:16:16.425
[info] System.getProperty("os.name") == 'Mac OS X'
[info] System.getProperty("os.version") == '10.8.5'
[info] System.getProperty("os.arch") == 'x86_64'
[info] System.getProperty("java.version") == '1.8.0_73'
[info] System.getProperty("java.vendor") == 'Oracle Corporation'
[info] System.getProperty("sun.arch.data.model") == '64'
[info] Enabling OpenGL.
[info] Biome profile activated: default
[info] using '.minecraft' directory at: '/Users/2frei/Library/Application Support/minecraft', libraries: '/Users/2frei/Library/Application Support/minecraft/libraries'
[info] using biome profiles at: 'biome'
[info] Scanning for profiles.
[info] Successfully loaded profile list.
[info] Beginning latest version list load.
[info] Attempting to download remote version list...
[info] Successfully loaded version list. URL: https://launchermeta.mojang.com/mc/game/version_manifest.json
[info] using minecraft launcher profile 'test5' with versionId '1.5.2'
[info] Loading libraries.
[info] Found library: /Users/2frei/Library/Application Support/minecraft/libraries/net/minecraft/launchwrapper/1.5/launchwrapper-1.5.jar
[info] Found library: /Users/2frei/Library/Application Support/minecraft/libraries/net/sf/jopt-simple/jopt-simple/4.5/jopt-simple-4.5.jar
[info] Found library: /Users/2frei/Library/Application Support/minecraft/libraries/org/ow2/asm/asm-all/4.1/asm-all-4.1.jar
[info] Found library: /Users/2frei/Library/Application Support/minecraft/libraries/net/java/jinput/jinput/2.0.5/jinput-2.0.5.jar
[info] Found library: /Users/2frei/Library/Application Support/minecraft/libraries/net/java/jutils/jutils/1.0.0/jutils-1.0.0.jar
[info] Skipping library: org.lwjgl.lwjgl:lwjgl:2.9.0
[info] Skipping library: org.lwjgl.lwjgl:lwjgl_util:2.9.0
[info] Skipping library: org.lwjgl.lwjgl:lwjgl-platform:2.9.0
[warning] Failed attempt to load library at: /Users/2frei/Library/Application Support/minecraft/libraries/org/lwjgl/lwjgl/lwjgl/2.9.1-nightly-20130708-debug3
[info] Skipping library: org.lwjgl.lwjgl:lwjgl:2.9.1-nightly-20130708-debug3
[warning] Failed attempt to load library at: /Users/2frei/Library/Application Support/minecraft/libraries/org/lwjgl/lwjgl/lwjgl_util/2.9.1-nightly-20130708-debug3
[info] Skipping library: org.lwjgl.lwjgl:lwjgl_util:2.9.1-nightly-20130708-debug3
[warning] Failed attempt to load library at: /Users/2frei/Library/Application Support/minecraft/libraries/org/lwjgl/lwjgl/lwjgl-platform/2.9.1-nightly-20130708-debug3
[info] Skipping library: org.lwjgl.lwjgl:lwjgl-platform:2.9.1-nightly-20130708-debug3
[info] Found library: /Users/2frei/Library/Application Support/minecraft/libraries/net/java/jinput/jinput-platform/2.0.5/jinput-platform-2.0.5-natives-osx.jar
[info] Found library: /Users/2frei/Library/Application Support/minecraft/libraries/net/java/jinput/jinput-platform/2.0.5/jinput-platform-2.0.5-natives-osx.jar
[crash] Amidst has encounted an uncaught exception on the thread Thread[Thread-6,1,main]
[crash] java.lang.NoClassDefFoundError: org/lwjgl/LWJGLException
	at java.lang.Class.getDeclaredFields0(Native Method)
	at java.lang.Class.privateGetDeclaredFields(Class.java:2583)
	at java.lang.Class.getDeclaredFields(Class.java:1916)
	at amidst.mojangapi.minecraftinterface.RecognisedVersion.getMainClassFields(RecognisedVersion.java:130)
	at amidst.mojangapi.minecraftinterface.RecognisedVersion.generateMagicString(RecognisedVersion.java:124)
	at amidst.mojangapi.minecraftinterface.RecognisedVersion.from(RecognisedVersion.java:119)
	at amidst.mojangapi.minecraftinterface.local.LocalMinecraftInterfaceBuilder.create(LocalMinecraftInterfaceBuilder.java:32)
	at amidst.mojangapi.file.directory.VersionDirectory.createLocalMinecraftInterface(VersionDirectory.java:99)
	at amidst.mojangapi.MojangApi.set(MojangApi.java:71)
	at amidst.gui.profileselect.LocalProfileComponent.tryLoad(LocalProfileComponent.java:84)
	at amidst.threading.WorkerExecutor.lambda$run$11(WorkerExecutor.java:71)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.ClassNotFoundException: org.lwjgl.LWJGLException
	at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	... 14 more
`

**Comment by stefandollase** [comment]
>> You always talk about a save game. Can you try to generate a random map, so we can see whether the issue also exists there?
> 
> No, each time I test a new MC version, I first generate a new world, and then I talk about it as the saved world … sorry to not be clear.

I guess I also did not express myself very clear: You can use Amidst without creating a world using Minecraft. To do this, please select the menu option *File* > *New from random seed* from the Amidst menu. You seem to be using *File* > *Open save game...*. Please try *File* > *New from random seed* and respond whether this also causes the issue.

As long as you did not touch the `.minecraft/versions/1.5.2/` directory you should be fine. Also, you should not touch the `.minecraft/libraries/` directory.

**Comment by 2frei** [comment]
Cannot select the 'New from random seed' menu because amidst crashes just after select my profile in 1.5.2 … but it do not with a profile in 1.6.4 or higher.

> These libraries should have been downloaded by the Minecraft launcher, because you ran the Minecraft profile.

I checked this and you're right, just the 2.9.0 version of lwjgl is present in my .minecraft/library directory … even though I totally erase and then properly DL MC.

> Also, you should not touch the .minecraft/libraries/ directory.

Too late ^^
I copied the 2.9.1 version of lwjgl (the 3 folders) from my MultiMC app (I insist MultiMC is installed on an other disk of my computer to do not interfere with MC Vanilla) into the .minecraft/library/org/lwjgl/lwjgl/ folder.
I renamed all the new folders and files : '2.9.1' by '2.9.1-nightly-20130708-debug3'
And then : Amidst works fine … at least, if I only use the 'New from random seed' menu, but I only use the 'New from random seed' menu ^^
I checked the obtained result and it is correct with my map.

I know this is not the approprite issue, but this is obvious that the problem comes from my launcher, so I will close this topic in many days (to let you time to answer) and hope that I will be the only one with this problem.

In any case, thanks for the help, because I really don't what is lwjgl and I will hereafter always check at the warning messages before asking help ^^

**Comment by stefandollase** [comment]
> Cannot select the 'New from random seed' menu because amidst crashes just after select my profile in 1.5.2 … but it do not with a profile in 1.6.4 or higher.

Well, yes, I did not get that. Thanks for the clarification :-)

I am glad you got this to work by yourself. Indeed, it often helps to read the log messages :-P

However, it seems that there is indeed an issue with Amidst misinterpreting the library description that is provided by the Minecraft launcher. I guess, Mojang recently changed the format of it: They are now able to provide different versions of a library for different versions of the operating system. For Minecraft 1.5.2, and OS X 10.5.*, they provide lwjgl 2.9.1. For Minecraft 1.5.2 and all other OS X versions, they provide lwjgl 2.9.0. Amidst currently misinterprets this, because it is not aware of the operating system version information. Thus, it wants to use lwjgl 2.9.1 for all OS X versions. However, your OS X version is 10.8.5, which means the Minecraft launcher should download and use lwjgl 2.9.0.

Here is an example of what I am talking about:

```
    {
      "name": "org.lwjgl.lwjgl:lwjgl:2.9.0",
      "rules": [
        {
          "action": "allow"
        },
        {
          "action": "disallow",
          "os": {
            "name": "osx",
            "version": "^10\\.5\\.\\d$"
          }
        }
      ],
      /* ... */
    },
    {
      "name": "org.lwjgl.lwjgl:lwjgl:2.9.1-nightly-20130708-debug3",
      "rules": [
        {
          "action": "allow",
          "os": {
            "name": "osx",
            "version": "^10\\.5\\.\\d$"
          }
        }
      ],
      /* ... */
    },
```

I won't promise when this will be fixed, but I hope to get this done next week.

**Comment by stefandollase** [comment]
This is fixed by #280.
